### PR TITLE
resolve issue 490

### PR DESCRIPTION
**Pull request by jamescheney**
Seems to fix the problem with polymorphic flat queries, but shredding/nested queries still need work.

@wricciot, over to you?

### CODE REVIEW

**Comment by jamescheney** [comment]
NB: The same approach (using a designated name for fields whose type is unit) also works for #567 at least in the case of flat queries.  For nested queries the errors encountered in #567 still occur so more work is needed.

**Comment by jamescheney** [comment]
If this now fixed #490 for both flat and nested queries then we should add tests similar to `empty.links` to the `shredding` tests directory.  (Including one that deals with an empty-list subexpression arising as a result of substituting in a function definition and normalizing, which was the original situation that @SimonJF encountered relating to #490 though it took us a little time to realize it.)

**Comment by jamescheney** [comment]
Wilmer and I just discussed and added some more interesting tests (for both flat and nested queries).  This should be ready to merge once the CI finishes.

**Comment by jamescheney** [comment]
This also fixes #567.

**Comment by jamescheney** [comment]
@slindley, could you take a quick look at this before merging?

File: core/query/evalNestedQuery.ml [hunk 1]
```diff
@@ -732,17 +732,29 @@ Avoiding unnecessary static indexes, or multiplexing pairs (a,d) where a is usua
...
+      | _ -> assert false
+    in
+    let idx_and_val = function
+      | `Record fl ->
+           let w_tmpl = try List.assoc "2" fl with NotFound _ -> `Record [] in
+           (match List.assoc "1" fl with
+            | `Record fl' -> (match List.assoc "1" fl', List.assoc "2" fl' with
+              | `Primitive (_,a_idx), `Primitive (_,d_idx) -> (a_idx,d_idx,w_tmpl)
+              | _ -> assert false)
+            | _ -> assert false)
```
Comment (slindley) [inline core/query/evalNestedQuery.ml]: I find `begin ... end` easier to read for nested pattern matching than `(...)`

**Comment by jamescheney** [comment]
I get what is going on now: in the shredding code we are assuming that all of the queries are returning (at least) some special index fields and Wilmer's changes workaround that.

I think this is symptomatic of the fact that the current interface boundary between ordinary evaluation and (shredded) query evaluation is at the wrong place: shredding returns a shredded package decorated with SQL query strings, and in some cases when we statically know the query result, but the result is not something that SQL can straightforwardly express, we should have the option of just calculating the (empty) result directly and not round-tripping to the database.  However, this workaround works for now and I think we should merge this and create an issue to record this requirement as part of the logner-term goal of refactoring the query subsystem to make this work better.
